File: leetcode/19-remove-nth-node-from-end-of-list/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

// naive approach
// iterate once for the length
// iterate one more time to remove the target node
// beats 100%
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	cnt := 0
	cur := head
	for cur != nil {
		cnt++
		cur = cur.Next
	}
	targetIdx := cnt - n
	dump := &ListNode{0, head}
	prev := dump
	cur = head
	cnt = 0
	for cur != nil {
		if targetIdx == cnt {
			prev.Next = cur.Next
			break
		}
		cnt++
		prev = prev.Next
		cur = cur.Next
	}
	return dump.Next
}

func main() {}

File: leetcode/19-remove-nth-node-from-end-of-list/main.py

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


"""
    1st approach: 2 pointers + dump
    - the gap is the n such that when fast reaches to the end, slow.next is the target

    e.g. 
    
    list = 1 -> 2 -> 3 -> 4 -> 5, target = 4
    
    1 -> 2 -> 3 -> 4 -> 5
              ^         ^
            slow        fast

    Time		O(n+target)
    Space       O(1)
    24 ms, faster than 72.46%
"""


class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        if n <= 0:
            return head
        dumphead = ListNode(0)
        dumphead.next = head
        slow = dumphead
        fast = dumphead
        for _ in range(n):
            fast = fast.next

        while fast.next != None:
            slow = slow.next
            fast = fast.next
        slow.next = slow.next.next
        return dumphead.next
